stint/intops.nim

```diff
@@ -7,155 +7,355 @@
-  result.data = high(type result.data)
+func `-`*(a, b: StInt): StInt =
+  ## Substraction for multi-precision unsigned int
+  result.imp.diff(a.imp, b.imp)
 
-func low*[bits](_: typedesc[StInt[bits]]): StInt[bits] {.inline.} =
-  result.data = low(type result.data)
-func low*[bits](_: typedesc[StUint[bits]]): StUint[bits] {.inline.} =
-  result.data = low(type result.data)
+func `-=`*(a: var StInt, b: StInt) =
+  ## In-place substraction for multi-precision unsigned int
+  a.imp.diff(a.imp, b.imp)
 
-import ./private/uint_exp, math
+func inc*(a: var StInt, w: Word = 1) =
+  a.imp.inc(w)
 
-func pow*(x: StUint, y: Natural): StUint {.inline.} =
-  ## Returns x raised at the power of y
-  when x.data is UintImpl:
-    result.data = x.data.pow(y)
+func `+`*(a: StInt, b: SomeUnsignedInt): StInt =
+  ## Addition for multi-precision unsigned int
+  ## with an unsigned integer
+  result.imp.sum(a.imp, Word(b))
+
+func `+=`*(a: var StInt, b: SomeUnsignedInt) =
+  ## In-place addition for multi-precision unsigned int
+  ## with an unsigned integer
+  a.imp.inc(Word(b))
+
+{.pop.}
+
+# Exponentiation
+# --------------------------------------------------------
+
+{.push raises: [], noinit, gcsafe.}
+
+func isOdd(x: Natural): bool =
+  bool(x and 1)
+
+func pow*(a: StInt, e: Natural): StInt =
+  ## Compute ``a`` to the power of ``e``,
+  ## ``e`` must be non-negative
+  if a.isNegative:
+    let base = a.neg
+    result.imp = base.imp.pow(e)
+    if e.isOdd:
+      result.negate
   else:
-    result.data = x.data ^ y
+    result.imp = a.imp.pow(e)
+
+func pow*[aBits, eBits](a: StInt[aBits], e: StInt[eBits]): StInt[aBits] =
+  ## Compute ``x`` to the power of ``y``,
+  ## ``x`` must be non-negative
+  doAssert e.isNegative.not, "exponent must be non-negative"
 
-func pow*(x: StUint, y: StUint): StUint {.inline.} =
-  ## Returns x raised at the power of y
-  when x.data is UintImpl:
-    result.data = x.data.pow(y.data)
+  if a.isNegative:
+    let base = a.neg
+    result.imp = base.imp.pow(e.imp)
+    if e.isOdd:
+      result.negate
   else:
-    result.data = x.data ^ y.data
-]#
\ No newline at end of file
+    result.imp = a.imp.pow(e.imp)
+
+{.pop.}
+
+# Division & Modulo
+# --------------------------------------------------------
+{.push raises: [], inline, noinit, gcsafe.}
+
+func `div`*(n, d: StInt): StInt =
+  ## Division operation for multi-precision unsigned uint
+  var tmp{.noinit.}: StInt
+
+  if n.isPositive:
+    if d.isPositive:
+      # pos / pos
+      result.imp = n.imp div d.imp
+      return
+    else:
+      # pos / neg
+      tmp = d.neg
+      result.imp = n.imp div tmp.imp
+      result.negate
+      return
+
+  let nneg = n.neg
+  if d.isNegative:
+    # neg / neg
+    tmp = d.neg
+    result.imp = nneg.imp div tmp.imp
+    return
+
+  # neg / pos
+  result.imp = nneg.imp div d.imp
+  result.negate
+
+func `mod`*(x, y: StInt): StInt =
+  ## Remainder operation for multi-precision unsigned uint
```

Comment:
The description should be clear of the sign of the remainder due to confusion between C and mathematical modulo/remainder.

Also whatever is chosen we should have `x = q*m+r` with q the quotient and r the remainder/modulo and m some multiple.